cmd/internal/cli/push.go

```diff
@@ -34,6 +45,32 @@ var PushCmd = &cobra.Command{
 	Run: func(cmd *cobra.Command, args []string) {
 		// Push to library requires a valid authToken
 		if authToken != "" {
+			if !unauthenticatedPush {
+				// check if the container is signed
+				imageSigned, err := signing.IsSigned(args[0], "https://keys.sylabs.io", 0, false, authToken, true)
+				if err != nil {
+					sylog.Warningf("%v", err)
```

Comment:
The `sylog.Warningf("%v", err)` will be a error like:

```
WARNING: unable to verify container: error while searching for signature blocks: no signatures found for system partition
```

After the error we ask the user if they want to continue using this unsigned container. Thats why its a `Warningf()` instead of a `Fatalf()`.

Reply:
Hmm... I would expect `signing.IsSigned` to return `true, nil`, `false, nil` or `false, non-nil`. The first two are cases where the code was able to do whatever it needs to do to figure out if the image is signed or not. The third option means it was not able to do its job because there was an error, so I would expect that at that point it's not possible to continue.

In other words, "this thing does not have signatures" should return `false, nil`, because that's what it means not being signed. If it _is_ signed, but we don't recognize the signatures, that's a different situation.

Maybe what this needs is _two_ functions: `IsSigned` and `IsSignatureValid`?

Reply:
:thinking: ... I originally did it like this is to provide a some debugging, like *`there was a error downloading the public key`* or something like that, the container *could* still be signed, the user will see that it was *`unable to verify the container...`*, with the possibility it *could* still be signed with a valid key.

Thats just my thoughts, I'm open to changing it like how you said...

Reply:
My take is that if the user is asking for strict signature verification (unauthenticatedPush == false) errors during signature verification are fatal, not just warnings.

Like I said before, I cannot imagine a situation where `signing.IsSigned` should return `true, non-nil`.

The possibilities are:

- true, nil: image is properly signed → push
- false, nil: image is not signed → don't push
- false, non-nil: signature could not be verified → don't push, bail out
- true, non-nil: ??? → don't push (and possibly freak out)

Reply:
This can return `true`, and an `error` if the container is signed, but you dont have a local key matching the signed container. Heres that that will look like:

```bash
$ singularity pull library://alpine:latest
INFO:    Downloading library image
 2.59 MiB / 2.59 MiB [============================================] 100.00% 440.96 KiB/s 6s
INFO:    Key with ID EDECE4F3F38D871E not found in local keyring, downloading from keystore...
INFO:    Container is signed
Data integrity checked, authentic and signed by:
	Sylabs Admin <support@sylabs.io>, Fingerprint 8883491F4268F173C6E5DC49EDECE4F3F38D871E
WARNING: no local key matching entity

$ echo $?
1
```

Although, if you try to push a broken container (I deleted some lines in the sif file) it will not bail out, _it probably should_:

```bash
$ singularity shell test.sif 
FATAL:   container creation failed: mount /proc/self/fd/3->/usr/local/var/singularity/mnt/session/rootfs error: can't mount image /proc/self/fd/3: failed to mount squashfs filesystem: invalid argument

$ singularity push test.sif library://westleyk/foo/bar:latest
WARNING: sylabs-token files are deprecated. Use 'singularity remote' to manage remote endpoints and tokens.
WARNING: No default remote in use, falling back to: https://library.sylabs.io
WARNING: unable to verify container: failed to parse signature block                   <- should bail out after this error message
INFO:    TIP: Learn how to sign your own containers here : https://www.sylabs.io/docs/

Your container is **NOT** signed! You REALLY should sign your container before pushing!
Do you really want to continue? [N/y] 
```